File: website/frontend/public/game/js/game_logic.js
// Source of truth: shared/game_constants.json
// Pure logic module — ZERO rendering/GUI code.

if (typeof require !== 'undefined') {
    var { Maze } = require('./maze');
    var { Pacman } = require('./pacman');
    var { Ghost } = require('./ghost');
}

const PELLET_POINTS = 10;

class GameLogic {
    constructor() {
        this._maze = new Maze();
        this._pacman = new Pacman();
        this._ghosts = [
            new Ghost(10, 10, 'Blinky', '#FF0000'),
            new Ghost(10, 11, 'Pinky', '#FFB8FF'),
            new Ghost(11, 10, 'Inky', '#00FFFF'),
            new Ghost(11, 11, 'Clyde', '#FF00FF'),
        ];
        this._score = 0;
        this._gameOver = false;
    }

    // ═══════════════════════════════════════════════════════
    // EDUCATIONAL NOTE: Game Loop Timing
    // ═══════════════════════════════════════════════════════
    // GameLogic.update() is the single "tick" function called once per
    // frame by the game loop in game.js. It orchestrates the update
    // order: Pacman moves first, then all ghosts, then collisions are
    // checked. This ordering matters — if ghosts moved first, Pacman
    // could "dodge through" a ghost that just vacated a tile.
    //
    // Compare with:
    //   Java → GameLogic.update() is called from a javax.swing.Timer's
    //          ActionListener at a fixed interval; same update ordering.
    //   Python → game_logic.update() is called from the pygame while-loop;
    //            Python's GIL ensures single-threaded execution just like
    //            the JS event loop and Java's EDT.
    //
    // KEY CONCEPT: This "logic-only" update method contains no rendering
    // or timing code. The game loop (game.js) controls WHEN to call it;
    // this class controls WHAT happens each tick. This separation is
    // consistent across all three language implementations.
    // ═══════════════════════════════════════════════════════

    /**
     * Main update tick: move pacman, move all ghosts, check collisions.
     * Does nothing if the game is already over.
     */
    update() {
        if (this._gameOver) return;

        this._pacman.update(this._maze);

        for (const ghost of this._ghosts) {
            ghost.update(this._maze);
        }

        this._checkCollisions();
    }

    // ═══════════════════════════════════════════════════════
    // EDUCATIONAL NOTE: Collision Detection
    // ═══════════════════════════════════════════════════════
    // Grid-based collision detection is the simplest possible approach:
    // two entities collide if and only if they occupy the same (x, y)
    // tile. No bounding boxes, no hitboxes, no pixel-perfect checks —
    // just integer equality. This works because all movement snaps to
    // the grid. Two collision types are checked here:
    //   1. Pacman vs. Pellet (same tile = collect, add score)
    //   2. Pacman vs. Ghost (same tile = game over)
    //
    // Compare with:
    //   Java → Identical integer comparison in _checkCollisions(); Java
    //          uses == for primitive int comparison (same semantics as
    //          JavaScript's === for integers).
    //   Python → Same logic in _check_collisions(); Python uses == which
    //            works correctly for integer comparison.
    //
    // KEY CONCEPT: JavaScript's === (strict equality) is used here for
    // integer grid coordinates. Unlike ==, it does not perform type
    // coercion, making it the safe choice. For grid-based games, this
    // O(ghosts) collision check is far simpler than AABB or circle
    // intersection tests needed in continuous-movement games.
    // ═══════════════════════════════════════════════════════

    /**
     * Checks two types of collisions:
     *   1. Pacman vs. Pellets — collect pellet, add score
     *   2. Pacman vs. Ghosts — game over
     */
    _checkCollisions() {
        const pacX = this._pacman.gridX;
        const pacY = this._pacman.gridY;

        // Pellet collection
        if (this._maze.hasPellet(pacX, pacY)) {
            this._maze.removePellet(pacX, pacY);
            this._score += PELLET_POINTS;
        }

        // Ghost collision — any ghost sharing Pacman's tile ends the game
        for (const ghost of this._ghosts) {
            if (pacX === ghost.gridX && pacY === ghost.gridY) {
                this._gameOver = true;
            }
        }
    }

    get maze() {
        return this._maze;
    }

    get pacman() {
        return this._pacman;
    }

    get ghosts() {
        return this._ghosts;
    }

    get score() {
        return this._score;
    }

    get gameOver() {
        return this._gameOver;
    }
}

if (typeof module !== 'undefined' && module.exports) {
    module.exports = { GameLogic };
} else if (typeof window !== 'undefined') {
    window.GameLogic = GameLogic;
}

File: python/game_logic.py
# Source of truth: shared/game_constants.json

from maze import Maze
from pacman import Pacman
from ghost import Ghost

# ═══════════════════════════════════════════════════════
# EDUCATIONAL NOTE: Collision Detection
# ═══════════════════════════════════════════════════════
# GameLogic is the SINGLE place where game rules are enforced:
# pellet collection and ghost-kills-Pacman. Keeping all rule
# checks in one class (instead of spreading them across Pacman
# and Ghost) makes the rules easy to find, read, and modify.
#
# Compare with:
#   Java (Swing) → Java often puts collision logic inside the
#   JPanel's actionPerformed() or a dedicated "GameEngine" class.
#   The pattern is the same, but Python needs no ActionListener
#   interface, no @Override, and no public/private keywords.
#
# KEY CONCEPT: Centralizing game rules in one class follows the
# Single Responsibility Principle and simplifies debugging.
# ═══════════════════════════════════════════════════════


class GameLogic:
    """Pure game-rule layer: owns the Maze, tracks score, detects collisions.

    ZERO rendering imports. The rendering layer reads state through
    the public properties and calls check_collisions() each frame.
    """

    # ═══════════════════════════════════════════════════════
    # EDUCATIONAL NOTE: Game Loop Timing
    # ═══════════════════════════════════════════════════════
    # GameLogic is called once per frame by game_panel.update().
    # It does NOT own the timing -- the main loop and pygame's
    # clock.tick(FPS) control how often check_collisions() runs.
    # This separation means GameLogic is easy to unit-test: just
    # call check_collisions() directly without needing a clock.
    #
    # Compare with:
    #   Java (Swing) → ActionListener.actionPerformed() is called
    #   by a javax.swing.Timer at a fixed interval. Python's
    #   approach (caller drives the timing) is more explicit and
    #   does not require event-listener registration boilerplate.
    #
    # KEY CONCEPT: Logic classes should not own timing -- let the
    # game loop call them, making the code testable and flexible.
    # ═══════════════════════════════════════════════════════

    def __init__(self):
        self._maze: Maze = Maze()
        self._score: int = 0
        self._game_over: bool = False

    # ------------------------------------------------------------------
    # Collision detection (grid-based)
    # ------------------------------------------------------------------

    def check_collisions(self, pacman: Pacman, ghosts: list[Ghost]) -> None:
        """Check pellet collection and ghost collision every frame.

        1. Pellet collection: if Pacman is on a pellet, remove it and
           add 10 points.
        2. Ghost collision: if Pacman shares a tile with any ghost,
           set game_over to True.
        """
        pacman_x = pacman.grid_x
        pacman_y = pacman.grid_y

        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Collision Detection
        # ═══════════════════════════════════════════════════════
        # Pellet collection is a single grid-coordinate check:
        # "is there a pellet at Pacman's current tile?" Because
        # movement is tile-based, there is no need for bounding-box
        # or circle-intersection math -- just compare integers.
        #
        # Compare with:
        #   Java (Swing) → if (maze.hasPellet(px, py)) { ... }
        #   Identical logic. Python's f-string score display and
        #   lack of semicolons make the code shorter, but the
        #   algorithm is exactly the same in both languages.
        #
        # KEY CONCEPT: Grid-based collision = integer equality check.
        # No floating-point rounding errors, no overlap calculations.
        # ═══════════════════════════════════════════════════════

        # Pellet collection
        if self._maze.has_pellet(pacman_x, pacman_y):
            self._maze.remove_pellet(pacman_x, pacman_y)
            self._score += 10

        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Collision Detection
        # ═══════════════════════════════════════════════════════
        # Ghost collision iterates over the ghost list using a
        # Python for-in loop. If ANY ghost shares Pacman's tile,
        # the game ends immediately. The list[Ghost] type hint
        # is optional documentation -- Python would run the same
        # code without it, unlike Java which requires List<Ghost>.
        #
        # Compare with:
        #   Java (Swing) → for (Ghost g : ghosts) { if (...) }
        #   Java's enhanced for-loop is similar, but requires the
        #   collection to implement Iterable<Ghost>. In Python,
        #   any iterable (list, tuple, generator) works in a for loop.
        #
        # KEY CONCEPT: Python's for-in loop works with ANY iterable,
        # not just specific collection types -- duck typing again.
        # ═══════════════════════════════════════════════════════

        # Ghost collision
        for ghost in ghosts:
            if pacman_x == ghost.grid_x and pacman_y == ghost.grid_y:
                self._game_over = True

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    # ═══════════════════════════════════════════════════════
    # EDUCATIONAL NOTE: Rendering/Drawing
    # ═══════════════════════════════════════════════════════
    # These properties expose game state to the renderer. The
    # renderer reads game_logic.score to display the HUD and
    # game_logic.game_over to show the "GAME OVER" overlay.
    # GameLogic itself has ZERO knowledge of pygame or pixels.
    #
    # Compare with:
    #   Java (Swing) → public int getScore() / public boolean
    #   isGameOver(). Java naming convention uses "is" prefix for
    #   booleans. Python just uses the property name directly:
    #   game_logic.game_over (reads naturally as a statement).
    #
    # KEY CONCEPT: Properties create a clean API boundary between
    # the game rules layer and the rendering/display layer.
    # ═══════════════════════════════════════════════════════

    @property
    def maze(self) -> Maze:
        return self._maze

    @property
    def score(self) -> int:
        return self._score

    @property
    def game_over(self) -> bool:
        return self._game_over
